### ISSUE TITLE
Ignition sometimes fails with `context cancelled` when fetching objects over http(s)

### ISSUE THREAD

**Issue by dgonyeo** [coreos/bugs#2435]
# Issue Report #

### Container Linux Version ###

 greater than or equal to Ignition v0.23.0 (CL 1722), probably

### Environment ###

All

### Expected Behavior ###

Ignition always successfully fetches http(s) resources, unless there's an issue with the server or a timeout is exceeded.

### Actual Behavior ###

Ignition will fail sometimes if the httpTotal timeout has been configured. Ex: fetching the config on EC2:

```
[    8.814120] ignition[401]: Ignition v0.25.0
[    8.828803] systemd[1]: Started Ignition (setup).
[    8.835590] ignition[401]: reading system config file "/usr/lib/ignition/base.ign"
[    8.843084] systemd[1]: Starting Ignition (disks)...
[    8.849728] ignition[401]: no config URL provided
[    8.854748] ignition[401]: reading system config file "/usr/lib/ignition/user.ign"
[    8.861626] ignition[401]: no config at "/usr/lib/ignition/user.ign"
[    8.868144] ignition[401]: GET http://169.254.169.254/2009-04-04/user-data: attempt #1
[    8.878161] ignition[401]: GET result: OK
[    8.884347] ignition[401]: failed to fetch config: context canceled
[    8.891405] ignition[401]: failed to acquire config: context canceled
[FAILED] Failed to start Ignition (disks).
```

### Reproduction Steps ###

1. Write out a local file `config.ign` containing [this](https://gist.github.com/dgonyeo/4f0b626f0ee3c090c964492fedb9c22f)
2. Host another file `config.ign` on localhost:80 containing [this](https://gist.github.com/dgonyeo/72308ee9b87c538ef39a6ad2a092ec76)
3. Ignition will now rarely fail. This can be seen by running Ignition repeatedly until this happens: `true; while [[ $? -eq 0 ]]; do ./bin/amd64/ignition -clear-cache -config-cache ./config-cache -log-to-stdout -oem file -stage files; done`
4. Ignition will fail more reliably if `time.Sleep(time.Second * 2)` is inserted above [this line](https://github.com/coreos/ignition/blob/313c86e149fd7321db2c98a1189dd3f51e61594d/internal/resource/url.go#L418)

### Other Information ###

This is likely to be due to the context to enforce timeouts being cancelled [before the http body has been read](https://github.com/coreos/ignition/blob/313c86e149fd7321db2c98a1189dd3f51e61594d/internal/resource/http.go#L218). A patch that doesn't cancel the context until after the body has been read appears to fix the problem.

I think happens inconsistently due to racing with the [body cancelling logic in net/http/client.go](https://golang.org/src/net/http/client.go?#L322)

Things I don't understand:
- ~Why this happens fetching configs on EC2, where I don't believe any timeouts are configured.~
- Why [this attempt at a minimal demonstration of this](https://gist.github.com/dgonyeo/8ca592d77e97eb8e7c25aff2d6c2221b) runs successfully instead of encountering an error.

**Comment by dgonyeo** [comment]
@arithx reminded me that there's a [default timeout when fetching configs](https://github.com/coreos/ignition/blob/aca97d03c4dfdc73ab51b4624bd1fbcbd9ae5e0f/internal/exec/engine.go#L120-L121), so it now makes sense to me that this gets hit when fetching configs on EC2. Yay.

**Comment by ajeddeloh** [comment]
It might be that there's some internal buffer that is handled after a cancelation point. Like you can canel transfer, but once its been transfered and buffered reads to the buffer are not affected by the cancelation.

**Comment by bgilbert** [comment]
Fixed in https://github.com/coreos/coreos-overlay/pull/3240.

**Comment by bgilbert** [comment]
Should be fixed in alpha 1786.0.1 and beta 1772.1.1, due shortly.  Stable has never been affected.
